File: dec01/Main.hs
{-# LANGUAGE GHC2024 #-}
{-# LANGUAGE TemplateHaskell #-}

import Data.FileEmbed (embedFile)

import Data.ByteString qualified as BS
import Data.ByteString.Char8 qualified as C8

input :: String
input = C8.unpack $(embedFile "input.txt")

parse :: String -> Int
parse = \case
  ('R' : xs) -> read xs
  ('L' : xs) -> -(read xs)
  _ -> error "Invalid"

wrapAdd x y = (x + y) `mod` 100

part1 :: Int
part1 =
  length
    . filter (== 0)
    . scanl wrapAdd 50
    . map parse
    . lines
    $ input

flatten x
  | x < 0 = replicate (abs x) (-1)
  | otherwise = replicate x 1

part2 :: Int =
  length
    . filter (== 0)
    . scanl wrapAdd 50
    . concatMap (flatten . parse)
    . lines
    $ input

main :: IO ()
main = print part2

File: dec01/src/main.rs
use std::io::Read;

fn parse(s: &str) -> i32 {
	let rest = &s[1..];
	match &s[0..1] {
		"R" => rest.parse::<i32>().unwrap(),
		"L" => -rest.parse::<i32>().unwrap(),
		_ => panic!("Invalid"),
	}
}

const fn wrap_add(x: i32, y: i32) -> i32 {
	(x + y).rem_euclid(100)
}

fn flatten(x: i32) -> impl Iterator<Item = i32> {
	std::iter::repeat_n(x.signum(), x.unsigned_abs() as usize)
}

fn part1(input: &str) -> usize {
	input
		.lines()
		.map(parse)
		.scan(50, |state, x| {
			*state = wrap_add(*state, x);
			Some(*state)
		})
		.filter(|&x| x == 0)
		.count()
}

fn part2(input: &str) -> usize {
	input
		.lines()
		.map(parse)
		.flat_map(flatten)
		.scan(50, |state, x| {
			*state = wrap_add(*state, x);
			Some(*state)
		})
		.filter(|&x| x == 0)
		.count()
}

fn main() {
	let input = {
		let mut buf = String::new();
		std::io::stdin().read_to_string(&mut buf).unwrap();
		buf
	};
	println!("Part 1: {}\nPart 2: {}", part1(&input), part2(&input));
}
